optimum/neuron/modeling.py

```diff
@@ -506,3 +511,136 @@ def forward(
+        + TEXT_GENERATION_EXAMPLE.format(
+            processor_class="AutoTokenizer",
+            model_class="NeuronModelForCausalLM",
+            checkpoint="gpt2",
+        )
+    )
+    def forward(
+        self,
+        input_ids,
+        cache_ids,
+        start_ids=None,
+        output_hidden_states=False,
+        output_attentions=False,
+        attention_mask=None,
+        return_dict=False,
+    ):
+        if output_hidden_states or output_attentions or attention_mask is not None:
+            warnings.warn(
+                "Warning: These arguments are not used by forward(): \
+                (output_hidden_states, output_attentions, attention_mask)"
+            )
+        # Evaluate the output logits, storing the current key and values at the indices specified by cache_ids
+        out_logits = self.model.forward(input_ids, cache_ids, start_ids)
+        out_logits = out_logits[:, None, :]
+        # Since we are using a static cache, we don't need to return past keys and values
+        if return_dict:
+            return ModelOutput([("logits", out_logits)])
+        return (out_logits,)
+
+    def prepare_inputs_for_generation(self, input_ids, past_key_values=None, **kwargs):
+        # Sanity checks
+        if past_key_values is not None:
+            raise ValueError("This model does not support dynamic key, value cache.")
+        batch_size, sequence_length = input_ids.shape
+        if batch_size != self.batch_size:
+            raise ValueError(
+                f"The specified batch_size ({batch_size}) does not match the model static batch size ({self.batch_size})"
+            )
+        if sequence_length > self.max_length:
+            raise ValueError(
+                f"The current sequence length ({sequence_length}) exceeds the model static sequence length ({self.max_length})"
+            )
+        # convert attention_mask to start_ids
+        attention_mask = None
+        start_ids = None
+        if "attention_mask" in kwargs:
+            attention_mask = kwargs["attention_mask"]
+
+        if attention_mask is not None:
+            _, start_ids = attention_mask.max(axis=1)
+
+        if self.cur_len > 0:
+            # Only pass the last tokens of each sample
+            input_ids = input_ids[:, -1:]
+            # Specifiy the single index at which the new keys and values need to be stored
+            cache_ids = torch.as_tensor([self.cur_len], dtype=torch.int32)
+        else:
+            # All tokens will be processed: initialize the cached with all intermediate keys and values
+            cache_ids = torch.arange(input_ids.shape[-1], dtype=torch.int32)
+
+        # Increment the current cache index
+        self.cur_len += input_ids.shape[-1]
+        model_inputs = {
+            "input_ids": input_ids,
+            "cache_ids": cache_ids,
+            "start_ids": start_ids,
+        }
+
+        return model_inputs
+
+    def can_generate(self):
```

Comment:
property or attribute?
```suggestion
    def can_generate(self) -> bool:
```

Reply:
This is actually a method of the PretrainedModel class that is called by the GenerationMixin.generate method.
Since we don't inherit from PretrainedModel we need to provide it explicitly.
This is a copy-paste of what is done in `onnxruntime` GenerationMixin.
I can definitely add the type annotation though.